app/cli/commandlineparser.cpp

```diff
@@ -0,0 +1,335 @@
+#include "commandlineparser.h"
+
+#include <QCommandLineParser>
+#include <QRegularExpression>
+
+#if defined(Q_OS_WIN)
+#include <qt_windows.h>
+#endif
+
+static bool inRange(int value, int min, int max)
+{
+    return value >= min && value <= max;
+}
+
+class CommandLineParser : public QCommandLineParser
+{
+public:
+    enum MessageType {
+        Info,
+        Error
+    };
+
+    void setupCommonOptions()
+    {
+        setSingleDashWordOptionMode(QCommandLineParser::ParseAsLongOptions);
+        addHelpOption();
+        addVersionOption();
+    }
+
+    void handleHelpAndVersionOptions()
+    {
+        if (isSet("help")) {
+            showInfo(helpText());
+        }
+        if (isSet("version")) {
+            showVersion();
+        }
+    }
+
+    void showMessage(QString message, MessageType type) const
+    {
+    #if defined(Q_OS_WIN)
+        UINT flags = MB_OK | MB_TOPMOST | MB_SETFOREGROUND;
+        flags |= (type == Info ? MB_ICONINFORMATION : MB_ICONERROR);
+        QString title = "Moonlight";
+        MessageBoxW(nullptr, reinterpret_cast<const wchar_t *>(message.utf16()),
+                    reinterpret_cast<const wchar_t *>(title.utf16()), flags);
+    #endif
+        fputs(qPrintable(message), type == Info ? stdout : stderr);
+    }
+
+    [[ noreturn ]] void showInfo(QString message) const
+    {
+        showMessage(message, Info);
+        exit(0);
```

Comment:
But, wouldn't that just try to exit from the event loop? Event loop which isn't running at this point.

I could write this without calling `exit()`, and just return to `main()`, but then the `showVersion()` would need to be shown differently as it calls the global `exit()` internally. Otherwise the message showing would behave inconsistently.